cvmfs/util/pipe.h

```diff
@@ -0,0 +1,271 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#ifndef CVMFS_UTIL_PIPE_H_
+#define CVMFS_UTIL_PIPE_H_
+
+#include <pthread.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+#include <cassert>
+#include <cerrno>
+
+#include "exception.h"
+#include "gtest/gtest_prod.h"
+#include "util/export.h"
+#include "util/pointer.h"
+#include "util/single_copy.h"
+
+#ifdef CVMFS_NAMESPACE_GUARD
+namespace CVMFS_NAMESPACE_GUARD {
+#endif
+
+/**
+ * Describes the functionality of a pipe
+ */
+enum PipeType {
+  kPipeThreadTerminator = 0,  // pipe only used to signal a thread to stop
+  kPipeWatchdog,
+  kPipeWatchdogSupervisor,
+  kPipeWatchdogPid,
+  kPipeDetachedChild,
+  kPipeTest,
+  kPipeDownloadJobs,
+  kPipeDownloadJobsResults
+};
+
+/**
+ * Common signals used by pipes
+ */
+enum PipeSignals {
+  kPipeTerminateSignal = 1
+};
+
+template <PipeType pipeType>
+class CVMFS_EXPORT Pipe : public SingleCopy {
+  FRIEND_TEST(T_Util, ManagedExecRunShell);
+  FRIEND_TEST(T_Util, ManagedExecExecuteBinaryDoubleFork);
+  FRIEND_TEST(T_Util, ManagedExecExecuteBinaryAsChild);
+
+ public:
+  /**
+   * A pipe is a simple asynchronous communication mechanism. It establishes
+   * a unidirectional communication link between two file descriptors. One of
+   * them is used to only write to the pipe, the other one only to read from it.
+   *
+   * This class is a simple wrapper around the handling of a "standard" pipe
+   * that uses system calls.
+   *
+   * @note PipeType as class template parameter should symbolize the
+   *       functionality of the specific type, independent of what variable
+   *       name it has
+   */
+  Pipe() {
+    int pipe_fd[2];
+    MakePipe(pipe_fd);
+    fd_read_ = pipe_fd[0];
+    fd_write_ = pipe_fd[1];
+  }
+
+  /**
+   * Destructor closes all valid file descriptors of the pipe
+   */
+  ~Pipe() {
+    if (fd_read_ != -1) {
```

Comment:
Let's simply call `Close()`. It does unnecessarily reset the `fd_read_`, `fd_write_` members but it saves us a few lines of code.